### PR TITLE
New hlsl xoroshiro

### PR DESCRIPTION
**Pull request by Przemog1**

### CODE REVIEW

**Comment by Przemog1** [comment]
Don't merge this pr before https://github.com/Devsh-Graphics-Programming/Nabla/pull/557

File: include/nbl/builtin/hlsl/random/xoroshiro.hlsl [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+{
+	uint2 xoroshiro64_state_advance(uint2 state)
+	{
+		state[1] ^= state[0];
+		state[0] = rotl(state[0], 26u) ^ state[1] ^ (state[1]<<9u); // a, b
+		state[1] = rotl(state[1], 13u); // c
+		
+		return state;
+	}
+}
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/random/xoroshiro.hlsl:36]: make it into a common case struct, remove the typedefs (store the state as a member in the base class)

File: include/nbl/builtin/hlsl/random/xoroshiro.hlsl [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+#define _NBL_BUILTIN_GLSL_RANDOM_XOROSHIRO_HLSL_INCLUDED_
+
+//#include <nbl/builtin/hlsl/math/functions.hlsl>
+#include <nbl/builtin/hlsl/cpp_compat/vector.hlsl>
+
+// TODO[Przemek]: include functions.hlsl instead
+uint32_t rotl(NBL_CONST_REF_ARG(uint32_t) x, NBL_CONST_REF_ARG(uint32_t) k)
+{
+   return (x<<k) | (x>>(32u-k));
+}
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/random/xoroshiro.hlsl:16]: start a `bit` header https://en.cppreference.com/w/cpp/header/bit

implement `rotl` same as in C++ STL except that there's no `constexpr` qualifier :(

also make `mpl/bit.hlsl` and implement a compile time version with templates and `nbl::hlsl::integral_constant`

we'll add to it later

File: include/nbl/builtin/hlsl/random/xoroshiro.hlsl (hunk 1 above)
Comment (Przemog1) [reply include/nbl/builtin/hlsl/random/xoroshiro.hlsl]: like that?
https://github.com/Devsh-Graphics-Programming/Nabla/pull/561/commits/72a5c39cdb64cecfeca4270a1c7ba0544f90af59

File: include/nbl/builtin/hlsl/bit.hlsl [hunk 3]
```diff
@@ -0,0 +1,55 @@
+#ifndef _NBL_BUILTIN_HLSL_BIT_INCLUDED_
+#define _NBL_BUILTIN_HLSL_BIT_INCLUDED_
+
+#include <nbl/builtin/hlsl/cpp_compat.hlsl>
+
+
+#ifndef __HLSL_VERSION
+#include <bit>
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/bit.hlsl:8]: I still need `using ` aliases between std::rotl and `nbl::hlsl` for C++

File: include/nbl/builtin/hlsl/bit.hlsl [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+#else
+namespace nbl
+{
+namespace hlsl
+{
+
+uint32_t rotl(NBL_CONST_REF_ARG(uint32_t) x, NBL_CONST_REF_ARG(uint32_t) s)
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/bit.hlsl:15]: its templated in C++, should be templated here too

Comment (devshgraphicsprogramming) [reply include/nbl/builtin/hlsl/bit.hlsl:15]: btw `s` is supposed to be `int32_t`

btw don't use `NBL...REF_ARG(T)` on types smaller than 64bit, just use `[const] T`

File: include/nbl/builtin/hlsl/bit.hlsl [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+{
+    const uint32_t N = 32u;
+    const uint32_t r = s % N;
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/bit.hlsl:18]: `r` needs to be `decltype(s)`

File: include/nbl/builtin/hlsl/bit.hlsl [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    
+    if(r == 0)
+        return 0;
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/bit.hlsl:21]: this special case is not needed if you make the `if (r>=0)` I think in HLSL a shift by sizeof(T)*8 is likely to wipe out the bit pattern so `| (x>> (N-r))` won't be UB and just `| 0`

File: include/nbl/builtin/hlsl/bit.hlsl [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        return (x << r) | (x >> (N - r));
+    }
+    else
+    {
+        return std::rotr(x, -r);
+    }
+}
+
+uint32_t rotr(NBL_CONST_REF_ARG(uint32_t) x, NBL_CONST_REF_ARG(uint32_t) s)
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/bit.hlsl:33]: all the comments as above for this

File: include/nbl/builtin/hlsl/random/xoroshiro.hlsl (hunk 1 above)
Comment (devshgraphicsprogramming) [reply include/nbl/builtin/hlsl/random/xoroshiro.hlsl]: you could have just inherited, but ok

File: include/nbl/builtin/hlsl/bit.hlsl (hunk 3 above)
Comment (Przemog1) [reply include/nbl/builtin/hlsl/bit.hlsl:7]: done using this:
https://github.com/Devsh-Graphics-Programming/Nabla/blob/new_hlsl_xoroshiro/include/nbl/builtin/hlsl/cpp_compat.hlsl#L12-L17

File: include/nbl/builtin/hlsl/bit.hlsl (hunk 5 above)
Comment (Przemog1) [reply include/nbl/builtin/hlsl/bit.hlsl]: done

File: include/nbl/builtin/hlsl/bit.hlsl (hunk 6 above)
Comment (Przemog1) [reply include/nbl/builtin/hlsl/bit.hlsl]: done

File: include/nbl/builtin/hlsl/bit.hlsl (hunk 7 above)
Comment (Przemog1) [reply include/nbl/builtin/hlsl/bit.hlsl]: done

File: include/nbl/builtin/hlsl/bit.hlsl (hunk 3 above)
Comment (devshgraphicsprogramming) [reply include/nbl/builtin/hlsl/bit.hlsl:7]: clever and nice

**Pull request merged**